### PR TITLE
feat: add `maxContentRangeChunkSize` option

### PR DESCRIPTION
**Pull request by ledyba**
Currently, this library tends to send entire file when `'acceptRanges': true,`

In this PR, I added `maxChunkSize` to `send.SendOptions` to limit the max chunk size.

Test case: https://hexe.net/2022/09/17/17:04:55/
[Souce code is here](https://github.com/fairy-rockets/the-gear-of-seasons/blob/0f75ca4db7bff84c8bc5c062f50fbddbc075d26c/server/src/controller/both/EntityController.ts#L64-L125)

I'm using firefox 132.0.

Request:

```
GET /entity/eb3467a204907295921f367ac8187d67 HTTP/2
Host: hexe.net
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:132.0) Gecko/20100101 Firefox/132.0
Accept: video/webm,video/ogg,video/*;q=0.9,application/ogg;q=0.7,audio/*;q=0.6,*/*;q=0.5
Accept-Language: ja
Range: bytes=0-
DNT: 1
Connection: keep-alive
Referer: https://hexe.net/2022/09/17/17:04:55/
Sec-Fetch-Dest: video
Sec-Fetch-Mode: no-cors
Sec-Fetch-Site: same-origin
Accept-Encoding: identity
Priority: u=4
TE: trailers
```

Response:

```
HTTP/2 206 
server: nginx/1.26.1
date: Sat, 16 Nov 2024 14:48:38 GMT
content-type: video/mp4
content-length: 65347058
accept-ranges: bytes
cache-control: public, max-age=0
last-modified: Sat, 17 Sep 2022 08:04:49 GMT
etag: W/"3e51df2-1834a7a8ef4"
content-range: bytes 0-65347057/65347058
content-disposition: inline; filename="eb3467a204907295921f367ac8187d67.original.mp4"
X-Firefox-Spdy: h2
```

I also checked in MS Edge 131.0.2903.51.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
  - This patch does not affect benchmark, and also, there is'nt [`npm run benchmark` command](https://github.com/fastify/send/blob/60c37c23b46fdbdf8d3d3494b3879423e73bff4b/package.json#L42-L49)
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -38,6 +38,11 @@ declare namespace send {
      */
     acceptRanges?: boolean | undefined;
 
+    /**
+     * Limit max response size when acceptRanges is true, defaults to the entire file size.
+     */
+    maxChunkSize?: number | undefined;
```
Comment (mcollina) [inline types/index.d.ts:44]: Can you please add a test for this type? We used tsd.

**Review by ledyba — COMMENTED** [review]
Thanks for the review. I added `tsd` tests for `maxChunkSize` and `acceptRanges`.

Could you take another look?

**Comment by Uzlopak** [comment]
Is a maxChunkSize not basically a highWatermark? Should we name it like that?

**Comment by ledyba** [comment]
I don't think that `maxChunkSize` is the best name. I'm not an English native, so I'd appreciate any advice..

File: test/send.1.test.js [hunk 2]
```diff
@@ -37,6 +37,18 @@ test('send(file, options)', function (t) {
         .expect(shouldNotHaveHeader('Content-Range', t))
         .expect(200, '123456789', err => t.error(err))
     })
+
+    t.test('should chunk size should be limited', function (t) {
```
Comment (Fdawgs) [inline test/send.1.test.js:41]:
```suggestion
    t.test('should limit chunk size', function (t) {
```

File: README.md [hunk 3]
```diff
@@ -49,6 +49,11 @@ Enable or disable accepting ranged requests, defaults to true.
 Disabling this will not send `Accept-Ranges` and ignore the contents
 of the `Range` request header.
 
+##### acceptRanges
+
+Specify maximum chunk size, defaults to send the entire file size.
+This will be used when `acceptRanges` is true.
+
```
Comment (Fdawgs) [inline README.md:56]: This needs to be renamed to `maxChunkSize` and moved to be after `maxAge`, as the options are sorted alphabetically.

File: test/utils.js [hunk 4]
```diff
@@ -9,6 +9,12 @@ module.exports.shouldNotHaveHeader = function shouldNotHaveHeader (header, t) {
   }
 }
 
+module.exports.shouldHaveHeader = function shouldNotHaveHeader (header, t) {
```
Comment (Fdawgs) [inline test/utils.js:12]:
```suggestion
module.exports.shouldHaveHeader = function shouldHaveHeader (header, t) {
```

File: README.md (hunk 3 above)
Comment (ledyba) [reply README.md]: Thanks, I also re-order the option in other files.

**Review by ledyba — COMMENTED** [review]
Thank you for the review and typo suggestion!

Could you take another look?

File: lib/send.js [hunk 5]
```diff
@@ -130,12 +130,17 @@ function normalizeOptions (options) {
...
+    ? Number(options.maxChunkSize)
+    : null
+
   const root = options.root
     ? resolve(options.root)
     : null
 
   return {
     acceptRanges,
+    maxChunkSize,
```
Comment (Fdawgs) [inline lib/send.js:143]: Sorry @ledyba, could you also move this to be in alphabetical order as well please?

**Comment by Uzlopak** [comment]
The name is not that nice. maxChunkSize is imho a `highWatermark` . `@fastify/send` only supports `bytes` for range. So it is de facto a `highWaterMark`

File: lib/send.js [hunk 6]
```diff
@@ -546,6 +551,9 @@ function sendFileDirectly (request, path, stat, options) {
 
           // Content-Range
           statusCode = 206
+          if (options.maxChunkSize) {
+            ranges[0].end = Math.min(ranges[0].end, ranges[0].start + options.maxChunkSize - 1)
```
Comment (Uzlopak) [inline lib/send.js:555]: It doesnt seem right.

**Review by Uzlopak — CHANGES_REQUESTED** [review]
preemptive blocker

**Review by climba03003 — DISMISSED** [review]
I don't see a problem here.
In your request headers `Range: bytes=0-` means send me a file starting with bytes 0 to the end of the file.
And the response is actually does what it is requested.

Partial content return should be controlled by the request (client), not the response (server) to limit.

File: lib/send.js (hunk 5 above)
Comment (ledyba) [reply lib/send.js]: Thanks, I renamed to `highWatarmark` and move it in alphabetical order.

File: lib/send.js (hunk 6 above)
Comment (ledyba) [reply lib/send.js]: There are some unit tests that success, so I think you are pointing out some edge cases that cause bug. If so, could you tell me the condition or something?

**Review by ledyba — COMMENTED** [review]
Thanks for the review. I renamed to `highWatermark`.

**Comment by ledyba** [comment]
@climba03003 Thanks for the comment.

Here is the latest rfc(9110).

- [14.2. Range](https://httpwg.org/specs/rfc9110.html#field.range).
- [15.3.7. 206 Partial Content](https://httpwg.org/specs/rfc9110.html#status.206)

Here is a from `15.3.7. 206 Partial Content`:

> A server that supports range requests ([Section 14](https://httpwg.org/specs/rfc9110.html#range.requests)) will usually attempt to satisfy all of the requested ranges, since sending less data will likely result in another client request for the remainder. However, a server might want to send only a subset of the data requested for reasons of its own, such as temporary unavailability, cache efficiency, load balancing, etc. Since a 206 response is self-descriptive, the client can still understand a response that only partially satisfies its range request.

So,  I think this PR's behavior is, at least, acceptable in RFC.

**Comment by ledyba** [comment]
Thanks for understanding and naming recommendations!

I choiced `maxReturnSize`, because I thought that `maxStreamSize` might be confusing or difficult to understand unless you know that the underlying implementation uses `fs.createReadStream`.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by gurgunday — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -113,6 +113,11 @@ Provide a max-age in milliseconds for HTTP caching, defaults to 0.
 This can also be a string accepted by the
 [ms](https://www.npmjs.org/package/ms#readme) module.
 
+##### maxReturnSize
+
+Specify maximum response content size, defaults to the entire file size.
```
Comment (Fdawgs) [inline README.md:118]:
```suggestion
Specify the maximum response content size, defaults to the entire file size.
```

**Review by Fdawgs — APPROVED** [review]

**Review by gurgunday — COMMENTED** [review]
I think this is a bug fix no? Shouldn't it be on by default?

**Comment by gurgunday** [comment]
Cc @climba03003

**Comment by climba03003** [comment]
> I think this is a bug fix no? Shouldn't it be on by default?

It is not a bugfix, but a feature request. By default, it should return as requested by the client.

**Comment by ledyba** [comment]
Hi. I'm the author. I think this is not a bug fix. (Please see the title: it starts with `feat:` )  

https://github.com/fastify/send/pull/90#issuecomment-2493304805
In a nutshell, even if the client requests all entire file range, the server can limit the response size. It seems allowed in RFC.
So, this feature is disabled by default.
It can be helpful for some rare cases, such as if the client buffers entire response body in the memory, for example.

**Comment by mcollina** [comment]
@Uzlopak can you take another look?

**Review by mcollina — APPROVED** [review]
lgtm

**Review by climba03003 — APPROVED** [review]

**Review by Uzlopak — APPROVED** [review]
I dont like the name "maxReturnSize", because it is the content Range so it should be maxContentRangeChunkSize or something like that. maxContentRangeSize would be not precise enough, because the size in content-range shows the whole size of the file and not the size of the sent chunk.

But I am not blocking either.

**Comment by gurgunday** [comment]
I actually agree, this name can be confusing...

**Comment by Uzlopak** [comment]
> There are only two hard things in Computer Science: cache invalidation and naming things.

-- Phil Karlton

**Comment by gurgunday** [comment]
I'd prefer "maxContentRangeChunkSize", it's super clear as a global parameter

**Comment by ledyba** [comment]
Thanks! I renamed to `maxContentRangeChunkSize` again.

To kick CI, it looks some maintainer have to allow to run the github action. I will wait.

**Review by climba03003 — APPROVED** [review]
SLGTM

**Comment by ledyba** [comment]
According to github, 

> Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.

image

Sorry, can some maintainer merge this PR? (I don't know this configuration is possible!)
